README.md
# The International Data Spaces (IDS) Information Model

The Information Model is an RDFS/OWL-ontology covering the fundamental concepts of the [International Data Spaces](https://www.internationaldataspaces.org) (IDS), i.e. the types of *digital contents* that are exchanged by *participants* by means of the IDS *infrastructure components*. The ontology and its documentation are published at [https://w3id.org/idsa/core](https://w3id.org/idsa/core). The model development is led by the [Fraunhofer Institutes for Applied Information Technology FIT](https://www.fit.fraunhofer.de/en) and [Intelligent Analysis and Information Systems IAIS](https://www.iais.fraunhofer.de/en) with support by members of the [International Data Spaces Association](https://www.internationaldataspaces.org/) in the context of the *Information Model* sub-working group (SWG4). The group is chaired by [Christoph Lange](https://github.com/clange/) ([Fraunhofer FIT](https://www.fit.fraunhofer.de/)) and [Sebastian Tramp](https://sebastian.tramp.name) ([eccenca GmbH](https://www.eccenca.com)). 

The model development is based on [GitHub](https://github.com/International-Data-Spaces-Association/InformationModel), following a [defined branching model](CONTRIBUTING.md). Contributions and community feedback are maintained via the GitHub [ticketing system](images/Issue_Lifecycle.jpg). The [release process](images/Release_process.jpg) is aligned with the International Data Spaces Association architecture
working group meetings, i.e. there are roughly 2 releases scheduled per year with intermediary updates to the development branch. The current release version is *4.1.0*, with the latest revision *4.1.0*. The Information Model and associated resources published on GitHub are available under the [Apache License 2.0](https://github.com/International-Data-Spaces-Association/InformationModel/blob/master/LICENSE).

## Authors and contributors

The Information Model was originally created by 

- Jaroslav Pullmann, [Fraunhofer FIT](https://www.fit.fraunhofer.de/en) now with [Stardog Union](http://www.stardog.com/) – succeeded by Dr. Christoph Lange (current co-chair of the information model sub-working group)
- Dr. Christian Mader, [Fraunhofer IAIS](https://www.iais.fraunhofer.de/en) and
- Dr. Sebastian Tramp, [eccenca GmbH](https://www.eccenca.com) (current co-chair of the information model sub-working group)

with significant contributions, comments and support by (in alphabetical order):

- Haydar Akyürek, [Fraunhofer FIT](https://www.fit.fraunhofer.de/en)
- Sebastian Bader, [Fraunhofer IAIS](https://www.iais.fraunhofer.de/en)
- Matthias Böckmann, [Fraunhofer IAIS](https://www.iais.fraunhofer.de/en) 
- Dr. Sandra Geisler, [Fraunhofer FIT](https://www.fit.fraunhofer.de/en)
- Dr. Anna Kasprzik, [TIB](https://www.tib.eu/en)
- Dr. Christoph Lange, [Fraunhofer FIT](https://www.fit.fraunhofer.de/en) (see above)
- Jörg Langkau, [nicos AG](http://www.nicos-ag.com)
- Johannes Lipp, [Fraunhofer FIT](https://www.fit.fraunhofer.de/en)
- Lina Molinas Comet, [Fraunhofer FIT](https://www.fit.fraunhofer.de/en)
- Andreas W. Müller, [Schaeffler](https://www.schaeffler.de/content.schaeffler.de/en/)
- Dr. Michael Theß, [Signal Cruncher GmbH](https://signal-cruncher.com/)

Likewise the authors would like to thank numerous colleagues at Fraunhofer, active participants of the [International Data Spaces Association](https://www.internationaldataspaces.org/en/), the members of the W3C [Dataset Exchange Group](https://www.w3.org/2017/dxwg/wiki/Main_Page) and [Data Market Austria](https://datamarket.at/en/) for illuminating conversations and support in shaping the Information Model.

## Directory layout

There are following top-level files and directories:

- `codes`: Code lists of enumerated *literal values* and unique, *singleton concept instances*. Code lists are the preferred way to encode and reference re-usable instances of a homogeneous value range.
- `docs`: Documentation of the latest release of the Information Model and its classes.
- `examples`: Example instances providing a guidance on recommended model usage and best practices.
- `images`: Figures supporting the understanding and documentation of the Information Model
- `metamodel`: Utility models related to annotation and processing of `model` concepts.
- `model`: Core model split into sub-model directories each serving a particular modeling aspect.
- `references`: Versioned sources of relevant 3rd party models used as a reference (informative or effective, i.e. via concept import)
- `taxonomies`: Hierarchical arrangement of concepts (classes) extending a base concept defined by the core `model`.
- `testing`: Test resources (e.g. RDFUnit) used by tools and the build process to ensure the syntactic validity and semantic consistence of the Information Model
- `utils` : Utility scripts used for Information Model development and testing
- `views`: Individual tailored views defined upon the overall model for a particular purpose, e.g. the selection of Glossary terms
- `Ontology.ttl`: Top-level metadata of the Information Model.

## Related resources

Additional resources are provided:

- Latest information about IDS and all related resources: The [IDS-G](https://github.com/International-Data-Spaces-Association/IDS-G)

- Releases are provided in this [GitHub repository](https://github.com/International-Data-Spaces-Association/InformationModel/releases/)

- Overview on IDS Messages, descriptions and attribute explanations: [Message Overview](https://w3id.org/idsa/ids-messages)

- Java representation of the Information Model, a mapping to a POJO domain model with no reference to graph or ontology constructs. 
Instances of the Information Model might such conveniently by created, validated and (de)serialized in a strongly-typed way using 
any common development environment without requiring in-depth knowledge of RDF and related Semantic Web standards. 
The Java Information Model is provided as a single JAR library via the [Maven Repository](https://maven.iais.fraunhofer.de/artifactory/eis-ids-public/)
operated by Fraunhofer [IAIS](https://www.iais.fraunhofer.de/); cf. the [documentation](https://jira.iais.fraunhofer.de/stash/projects/ICTSL/repos/ids-infomodel-demo/):

```xml
<repositories>
    <repository>
        <id>snapshots</id>
        <name>maven-snapshots</name>
        <url>https://maven.iais.fraunhofer.de/artifactory/eis-ids-public</url>
    </repository>
</repositories>
```

The dependency to the Java representation is as follows (replace <code>${infomodel.version}</code> with the latest version):
```xml
<dependency>
    <groupId>de.fraunhofer.iais.eis.ids.infomodel</groupId>
    <artifactId>java</artifactId>
    <version>${infomodel.version}</version>
</dependency>
```

For automated validation of the Information Model POJOs (via the builder classes) you can include:
```xml
<dependency>
    <groupId>de.fraunhofer.iais.eis.ids.infomodel</groupId>
    <artifactId>validation-serialization-provider</artifactId>
    <version>${infomodel.version}</version>
</dependency>
```

- UML-based visualization of the Information Model, its classes, properties and relations, including their cardinalities. It is available as part of the Nightly and Release Packages on [GitHub](https://github.com/International-Data-Spaces-Association/InformationModel/releases).

- A high-level conceptual specification is given as a part of the IDS Reference Architecture Model ([current version 3.0](https://www.internationaldataspaces.org/publications/reference-architecture-model-3-0/) of 2019; see section 3.4).

- [Webinar of November 2018, giving a technical overview of the Information Model](https://www.internationaldataspaces.org/wp-content/uploads/2019/07/IDS-Information-Model.pdf)

For members of the International Data Spaces Association only:
- [Working document](https://industrialdataspace.jiveon.com/docs/DOC-1817), explaining
the usage of the Information Model for data exchanges in the IDS.


CHANGELOG.md

# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).

## [4.2.0] 2022-05-13
Version 4.2.0 of the IDS Information Model 

### Added

* AppRoute now subclass of ids:Described
* Individuals for quality metrics and a dimension: `idsc:AVAILABILITY and idsc:DOWNLOAD_URL_AVAILABILITY_METRIC`
* Class `ids:PaymentModality` to document pricing modalities (free/ fixed price/ negotiation basis).
* Class `ids:IdsProtocolSpecification` to describe `ids:Endpoint`s with their supported protocols
* New individuals of `ids:Frequency`: `TWO_TIMES_A_DAY`, `EVERY_THREE_HOURS`, `BIHOURLY`, `HOURLY`, `EVERY_30_MINUTES`, `EVERY_15_MINUTES`, `EVERY_10_MINUTES`, `EVERY_5_MINUTES`, `EVERY_1_MINUTE`
* New individuals of `ids:IdsProtocolSpecification`: `IDS-REST`, `IDS-CP`, `QUERY`, `MULTIPART`
* Shapes for AppResources
* Class UsageControlComponent and Shape

## [4.1.0] 2021-05-26
Version 4.1.0 of the IDS Information Model

### Added

* `ids:ConfigurationManager` can refer to the configured App Resources, Brokers, App Stores and Clearing Houses represented as catalogs.
* `ids:AppRoute` can refer to Broker and Clearing House instances (via `ids:appRouteClearingHouse` and `ids:appRouteBroker`), which are used for publishing and logging related communication.
* `ids:GenericEndpoint` for non-IDS endpoints.
* Catalog subclasses for Broker, AppStore, AppResource and ClearingHouse instances. Check the [Catalog.ttl](model/infrastructure/Catalog.ttl) class for details.
* Additional properties for Participants for identification and classification, e.g., legal name, business identifier, legal form and VAT ID (addressing requirements to participants in Gaia-X)
* Additional properties for `ids:DataApp` and `ids:AppRepresentation` to express supported usage policies as well as app-related runtime, configuration and deployment information.

### Changed:
* `ids:ParticipantRequestMessage` and `ids:ParticipantResponseMessage` are deprecated and will be removed in the next release. Using `ids:DescriptionRequestMessage` and `ids:DescriptionReponseMessage` is advised.


### Removed
* `idsm:` metamodel annotations, which were used to annotate cardinalities of properties. The property cardinalities and restrictions are now exclusively represented via the corresponding SHACL shapes found in the [testing subdirectory](testing/)


## [4.0.0] 2020-08-04
Version 4.0.0 of the IDS Information Model

### Added

* `ids:ConfigurationModel` class for Connector configuration and deployment related information.
  * Several complementary classes, including `ids:Proxy`, `ids:LogLevel`, `ids:AppRoute`, `ids:UserAuthentication`.
* Redesign of `ids:DataApp` class.
  * `ids:DataApp` as superclass of `ids:OrchestrationApp`, `ids:SmartDataApp`, `ids:SystemAdapterApp` .
  * New properties `ids:appDocumentation` with range `xsd:string` and `ids:appEndpoint` with range `ids:AppEndpoint` to enable app and app endpoint documentation.
* `ids:AppEndpoint`, a subclass of `ids:Endpoint` to describe data app endpoints, such as path suffixes, ports and endpoint documentation.
* Languages for `ids:Representation` via the `ids:language` property.
* `ids:ResourceCatalog`, `ids:ConnectorCatalog` and `ids:ParticipantCatalog` as subclasses of `ids:Catalog` to enable infrastructure components, e.g., Broker and ParIS, to present Connector- and Participant-specific catalogs.
* New messages:
  * Messages for app-related communication between App Provider and App Store.
  * `ids:UploadMessage` and `ids:UploadResponseMessage` to allow data upload to a recipient.

### Changed

* Major changes on the IDS Usage Policy Language. See section [IDS Usage Policy Language – Updates and major Changes](#usagepolicy-4.0.0) below.

* `ids:MediaType`. Removed instances for `ids:IANAMediaType`.
    * Media types via `ids:mediaType` with range `ids:IANAMediaType` should be used with the corresponding [IANA URLs](https://www.iana.org/assignments/media-types/media-types.xhtml), as these have complete coverage. E.g., for JSON-LD:
    ```JSON
    "ids:mediaType": {
      "@id":   "https://www.iana.org/assignments/media-types/application/ld+json",
      "@type": "ids:IANAMediaType"
    }
    ```
    * For custom media types (`ids:CustomMediaType`), it is advised to use a recognizable unique identifier.

* Renamed property `ids:catalog` of Connector to `ids:resourceCatalog` and changed range to `ids:ResourceCatalog`.
* Connectors can now present multiple catalogs via the `ids:resourceCatalog` property.
* Switched range of `ids:standardLicense` property from `ids:License` to `xsd:anyURI`. There is no vocabulary with a complete, regularly updated list of software and data licenses which we can provide as instances of the `ids:License` class, to the best of our knowledge. Users should provide the IRI of the correct license. We recommend using sources such as [Wikidata](http://www.wikidata.org/) to search for most common licenses.
* `ids:contentStandard` properties of class `ids:DigitalContent` and `ids:representationStandard` properties of class `ids:Representation` now have range `xsd:anyURI`.
* Redesign of `ids:Endpoint` class.
  * Removed `ids:InteractiveEndpoint`.
  * `ids:StaticEndpoint` is now called `ids:ConnectorEndpoint`. `ids:ConnectorEndpoint` contains additional properties, especially for endpoint description and documentation.
  * `ids:Host` information is now part of the `ids:Endpoint`.
  * New property: `ids:endpointInformation` with range `xsd:string` and `ids:endpointDocumentation` with range `xsd:anyURI` for endpoint description and documentation.
* Core classes for digital content (`ids:Resource`, `ids:DigitalContent`, `ids:Representation`, `ids:Artifact`) now aligned with W3C [DCAT 2](https://www.w3.org/TR/vocab-dcat-2/)

### Removed

create-ontology-ttl.sh
#!/bin/bash

write_to_file()
{

    # initialize a local var
    local file="Ontology.ttl"
    local version=$1
    
    if [[ -z "$version" ]] ; then
        echo Warning! No version parameter supplied. >&2
		echo Add version parameter to script. >&2
		echo Example: $'\n\t'sh create-ontology-ttl.sh 4.1.0 $'\n'>&2
		echo Exiting >&2
		exit 1;
    fi
    
    
    # delete old content and add prefixes
    cat > "$file" <<'EOF'
@prefix owl:  <http://www.w3.org/2002/07/owl#> .
@prefix dct:  <http://purl.org/dc/terms/> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix prov: <http://www.w3.org/ns/prov#> .
@prefix rdf:  <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix ids:  <https://w3id.org/idsa/core/> . 
@prefix vann: <http://purl.org/vocab/vann/> .
@prefix void: <http://rdfs.org/ns/void#> .
@prefix voaf: <http://purl.org/vocommons/voaf#> .
@prefix foaf: <http://xmlns.com/foaf/0.1/> .
@prefix cc:   <http://creativecommons.org/ns#> .
@prefix xsd:  <http://www.w3.org/2001/XMLSchema#>.

EOF

    # add meta-data
    
    cat >> "$file" <<EOF
# Description of this ontology
# ----------------------------

ids:
    a voaf:Vocabulary, owl:Ontology ;
    rdfs:isDefinedBy <https://w3id.org/idsa/core/> ;
    rdfs:label "IDS Information Model"@en ;
    dct:title "The International Data Spaces Information Model"@en ;
    rdfs:comment "This ontology defines classes and properties for describing participants, infrastructure, data and services of the International Data Spaces (formerly known as Industrial Data Space)." ;
    cc:license <http://www.apache.org/licenses/LICENSE-2.0> ;
    dct:creator
       <https://github.com/cmader> ,
       <https://github.com/jpullmann> ,
       <http://sebastian.tramp.name> ;
    dct:contributor
       _:AndreasMueller ,
       _:AnnaKasprzik ,
       <https://github.com/sebbader> ,
       <https://github.com/Maboeckmann> ,
       <https://github.com/clange> ,
       <https://github.com/HaydarAk> ,
       <https://github.com/jlangkau> ,
       <https://github.com/JohannesLipp> ,
       <https://github.com/mkollenstart> ,
       <https://github.com/lcomet> ;
    dct:publisher ids:IDSA ;
    dct:created "2017-09-26"^^xsd:date ;
    dct:modified "$(date +%Y-%m-%d)"^^xsd:date ;
    owl:versionInfo "${version}" ;
    owl:versionIRI <https://w3id.org/idsa/core/${version}> ;
    vann:preferredNamespaceUri "https://w3id.org/idsa/core/" ;
    vann:preferredNamespacePrefix "ids" ;
    rdfs:seeAlso <https://international-data-spaces-association.github.io/InformationModel/> ;
    void:vocabulary
       vann: ,
       void: ,
       voaf: ,
       dct: ,
       foaf: ;
.

<http://www.apache.org/licenses/LICENSE-2.0>
    rdfs:label "Apache License Version 2.0, January 2004" ;
.

ids:IDSA
    a dct:Agent, foaf:Organization ;
    foaf:homepage <https://www.internationaldataspaces.org> ;
    rdfs:label "International Data Spaces Association" ;
.

<http://sebastian.tramp.name>
    a dct:Agent, foaf:Person ;
    foaf:name "Sebastian Tramp" ;
.
<https://github.com/cmader>
    a dct:Agent, foaf:Person ;
    foaf:name "Christian Mader" ;
.
<https://github.com/jpullmann>
    a dct:Agent, foaf:Person ;
    foaf:name "Jaroslav Pullmann" ;
.
_:AndreasMueller
    a dct:Agent, foaf:Person ;
    foaf:name "Andreas Müller"
.
_:AnnaKasprzik a dct:Agent, foaf:Person ;
    foaf:name "Anna Kasprzik" ;
.
<https://github.com/sebbader> a dct:Agent, foaf:Person ;
    foaf:name "Sebastian Bader" ;
.
<https://github.com/HaydarAk> a dct:Agent, foaf:Person ;
    foaf:name "Haydar Akyürek" ;
.
<https://github.com/JohannesLipp> a dct:Agent, foaf:Person ;
    foaf:name "Johannes Lipp" ;
.
<https://github.com/clange> a dct:Agent, foaf:Person ;
    foaf:name "Christoph Lange" ;
.
<https://github.com/Maboeckmann> a dct:Agent, foaf:Person ;
    foaf:name "Matthias Böckmann" ;
.
<https://github.com/jlangkau> a dct:Agent, foaf:Person ;
    foaf:name "Jörg Langkau" ;
.
<https://github.com/mkollenstart> a dct:Agent, foaf:Person ;
    foaf:name "Maarten Kollenstart" ;
.
<https://github.com/JohannesLipp> a dct:Agent, foaf:Person ;
    foaf:name "Johannes Lipp" ;
.
<https://github.com/lcomet> a dct:Agent, foaf:Person ;
    foaf:name "Lina Molinas Comet" ;
.
EOF
    
    # search for files in selected folders
    echo '# ----------------------------' >> "$file"
    echo '# Imports of class files' >> "$file"
    echo 'ids:' >> "$file"

    # search for files in selected folders (omitting metamodel/)
    for class in $(find \
                       model/* \
                       taxonomies/ \
                       codes/ \
                       -maxdepth 1 -name "*.ttl")
    do
        if [[ -f $class ]]; then
            echo "    owl:imports <$class> ;" >> "$file"
        fi
    done
    
    echo ". " >> "$file"
}

# execute it
write_to_file $1
